docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,408 @@
+
+    // ipBlock defines policy on a particular IPBlock. If this field is set then
+    // neither of the other fields can be.
+    // +optional
+    IPBlock *networkingv1.IPBlock `json:"ipBlock,omitempty" protobuf:"bytes,3,rep,name=ipBlock"`
+}
+
+// Status defines the observed state of NetworkQoS
+type Status struct {
+    // A concise indication of whether the EgressNetworkQoS resource is applied with success.
+    // +optional
+    Status string `json:"status,omitempty"`
+
+    // An array of condition objects indicating details about status of EgressNetworkQoS object.
+    // +optional
+    // +patchMergeKey=type
+    // +patchStrategy=merge
+    // +listType=map
+    // +listMapKey=type
+    Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+```
+
+* Ensure backward compatibility support available for `EgressQoS` API.
+
+### Implementation Details
+
+The new controller is introduced in OVN-Kubernetes which would watch `NetworkQoS` in addition to `EgressQoS`, `Pod` and `Node` objects, which will create the relevant NetworkQoS objects and attach them to all of the node local switches in the cluster in OVN - resulting in the necessary flows to be programmed in OVS.
+
+In order to not create an OVN NetworkQoS object per pod in the namespace, the controller will also manage AddressSets.
+For each QoS rule specified in a given `NetworkQoS` it'll create an AddressSet, adding only the pods whose label matches the PodSelector to it, making sure that new/updated/deleted matching pods are also added/updated/deleted accordingly. Rules that do not have a PodSelector will leverage the namespace's AddressSet.
+
+Similarly when `NetworkQoS` is created for Pods secondary network, OVNK must create a new AddressSet for every QoS rule. When no pod selector is specified, then it must contain all of the pod's IP addresses that belong to the namespace and selected network. If only a set of pods are chosen via podSelector, then it must have IP addresses only for chosen pod(s).
+
+For example, using LGW mode and assuming there's a single node `node1` and the following `NetworkQoS` is created:
+
+```yaml
+kind: NetworkQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: qos-external-foo
+  namespace: default
+spec:
+  podSelector:
+    matchLabels:
+      priority: Critical
+  egress:
+  - priority: 100
+    dscp: 59
+    classifier:
+      ports:
+        protocol: "tcp"
+        port: 22
+      to:
+      - cidr: 1.2.3.4/32
+  - priority: 10
+    dscp: 46
+    classifier:
+      to:
+      - cidr: 1.2.3.4/32
+---
+kind: NetworkQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: qos-external-bar
+  namespace: default
+spec:
+  egress:
+  - dscp: 30
+    classifier:
+      to:
+      - cidr: 0.0.0.0/0
+```
+the equivalent of:
+```bash
+ovn-nbctl qos-add node1 to-lport 1000 "ip4.src == <default_ns-qos-external-foo address set> && ip4.dst == 1.2.3.4/32 && tcp && tcp.dst == 22" dscp=59
```

Comment:
Please use priority:

Ref: qos-add SWITCH DIRECTION PRIORITY MATCH [rate=RATE [burst=BURST]] [dscp=DSCP] [mark=MARK]\
https://github.com/ovn-org/ovn/blob/94aad069df918cbe7be800abc0790450931863ac/utilities/ovn-nbctl.c#L286

```
ovn-nbctl qos-add node1 to-lport 100 "ip4.src == <default_ns-qos-external-foo address set> && ip4.dst == 1.2.3.4/32 && tcp && tcp.dst == 22" dscp=59
ovn-nbctl qos-add node1 to-lport 10 "ip4.src == <default_ns-qos-external-foo address set> && ip4.dst == 1.2.3.4/32" dscp=46
ovn-nbctl qos-add node1 to-lport <qos-external-bar-priority-goes-here> "ip4.src == <default_ns address set> && ip4.dst == 0.0.0.0/0" dscp=30
```

Reply:
good catch, updated the commands.